application/plugin_base/item_list.py

```diff
@@ -1655,11 +1673,34 @@ def get_provider(self, path=None):
 		if path is None:
 			return self._current_provider
 
+		# create a local provider if no provider exists yet as fallback in case we are opening an archive
+		if len(self._providers) == 0:
+
+			Provider = self._parent.get_provider_by_protocol('file')
+			result = Provider(self, user.home)
+
+			# cache provider for later use
+			root_path = result.get_root_path(user.home)
+			self._providers[root_path] = result
+			self._current_provider = result
+
 		# check if there is a provider for specified path
 		if path in self._providers:
 			result = self._providers[path]
 
 		else:
+
+			# assuming a path like /home/user/some.zip/subdir1/subdir2
+			# creates intermediate archive provider if necessary
+			p = path
+			while p not in self._providers and p != '/' and '://' not in path:
```

Comment:
Hi Mladen

I have made another check-in which should simplify things a bit - or at
least make it easier to reason about the change:

1) in get_provider() - only new code is now triggered if the provider
selected can't handle the path requested e.g. path is "/home/user/some.zip"
but existing code would pick LocalProvider for /

2) in create_provider() - a default provider for user.home is created here
and the zip path is a little further processed here now

Re: this loop:

            # assuming paths like /home/user/some.zip or
/home/user/some.zip/subdir1/subdir2
            # bit by by bit, cut path back to the path of the archive itself
            p = path
            while p not in self._providers and p != '/' and '://' not in
path:
                mime_type =
self._parent.associations_manager.get_mime_type(path=p)
                if self._parent.is_archive_supported(mime_type):
                    path = p
                    break
                else:
                    p = os.path.dirname(p)


Currently a ZipProvider is only created with a path like
"/home/user/some.zip", but if one panel has a zip file open and
Ctrl-Left/Right is pressed on a subdir, the path requested on the other
panel is more like "/home/user/some.zip/subdir1/subdir2". The while loop
cuts off directory by directory until it hits the portion of the path
representing the zip file  in order to create a ZipProvider for it.

Test cases: user should be able to press Ctrl-Left/Right on a directory on
panels on both sides when

- both panels have "normal" LocalProvider
- on panel has "normal" LocalProvider, one has ZipProvider
- both panels have ZipProvider but for different archives
- additionally, any saved paths/providers from Sunflower was closed last
should be restored properly

I admit all a bit fiddly and some automated GUI testing would be nice to
capture use cases.

Cheers
Christian


On Sat, Feb 21, 2015 at 4:31 PM, Mladen Mijatov <notifications@github.com>
wrote:

> In application/plugin_base/item_list.py
> <https://github.com/MeanEYE/Sunflower/pull/65#discussion_r25123314>:
>
> >  		# check if there is a provider for specified path
> >  		if path in self._providers:
> >  			result = self._providers[path]
> >
> >  		else:
> > +
> > +			# assuming a path like /home/user/some.zip/subdir1/subdir2
> > +			# creates intermediate archive provider if necessary
> > +			p = path
> > +			while p not in self._providers and p != '/' and '://' not in path:
>
> What is this checking for?
>
> —
> Reply to this email directly or view it on GitHub
> <https://github.com/MeanEYE/Sunflower/pull/65/files#r25123314>.
>